### PR TITLE
Enforce EB sales limit.

### PR DESCRIPTION
**Pull request by umgelurgel**
It's a draft solution - haven't tested it in detail yet. It switches off EB fares when enough orders with EB tickets have been created; since we create orders before they're completed, very likely less tickets have been successfully sold when the limit triggers.

Keen on hearing feedback if this is worth pursuing.

### CODE REVIEW

File: conference/cart.py [hunk 1]
```diff
@@ -267,6 +269,18 @@ def send_order_confirmation_email(order: Order, current_site) -> None:
     )
 
 
+def handle_early_bird_ticket_limit():
```
Comment (artcz) [inline conference/cart.py]: I like the idea, I'm just wondering if it might be even simpler – calling this handler on step2's POST, and instead of disabling in db it would just throw an error that early birds are gone already. WDYT?

File: conference/fares.py [hunk 2]
```diff
@@ -191,6 +194,19 @@ def set_early_bird_fare_dates(conference, start_date, end_date):
...
 
+def disable_early_bird_fares(conference):
+    early_birds = Fare.objects.filter(
+        conference=conference,
+        code__regex=FARE_CODE_REGEXES["types"][FARE_CODE_TYPES.EARLY_BIRD],
+    )
+    assert (
+        early_birds.count()
+        == len(FARE_CODE_VARIANTS) * len(FARE_CODE_GROUPS)
+        == 3 * 5
```
Comment (artcz) [inline conference/fares.py]: the 3 * 5 part of the assertion is, for various reasons, likely to fail on production.
yes, it's perfect for dev, and dev is modeled after production, but it's more likely going to be 3*3 :)

File: conference/cart.py (hunk 1 above)
Comment (umgelurgel) [reply conference/cart.py]: Yeah, that sounds like an even better idea.

File: conference/cart.py [hunk 3]
```diff
@@ -267,6 +266,17 @@ def send_order_confirmation_email(order: Order, current_site) -> None:
...
 
 
+def is_early_bird_sold_out():
+    eb_ticket_orders = Ticket.objects.filter(
+        fare__conference=settings.CONFERENCE_CONFERENCE,
+        frozen=False,
+        # orderitem__order___complete=True,
+        fare__code__regex=FARE_CODE_REGEXES["types"][FARE_CODE_TYPES.EARLY_BIRD]
+    )
+
```
Comment (malemburg) [inline conference/cart.py]: This should probably also take the fare end dates into account, unless this is already done elsewhere.

File: templates/conference/cart/step_2_pick_tickets.html [hunk 4]
```diff
@@ -51,7 +51,7 @@ <h4>{{ fare.name }}</h4>
                         </td>
                     </tr>
                     {% empty %}
-                    No tickets available
+                    No tickets available - if you're looking for early bird tickets, it means they have sold out.
```
Comment (malemburg) [inline templates/conference/cart/step_2_pick_tickets.html]: I'd reword this to the more generic: "No tickets available, sorry. This could mean that ticket sales in the current category have not yet started or the tickets are already sold out (early bird tickets tend to sell out quickly) ".

File: pycon/settings.py [hunk 5]
```diff
@@ -469,6 +469,7 @@ def _(x):
     "ep2019",
     "ep2020",
 )
+EARLY_BIRD_ORDER_LIMIT = config("EARLY_BIRD_ORDER_LIMIT", default=210, cast=int)
```
Comment (malemburg) [inline pycon/settings.py]: I'd raise the default to 220, since experience shows that quite a few people leave tickets in the cart without actually ordering. The limit doesn't have to be a hard limit, only prevent going way beyond the limit.

**Review by malemburg — APPROVED** [review]

**Review by aramboi — APPROVED** [review]
Looks good. :shipit:

File: conference/cart.py (hunk 3 above)
Comment (matrixise) [reply conference/cart.py]: @malemburg currently, `fare__conference` will use the assigned fares for the current conference.

File: conference/cart.py [hunk 6]
```diff
@@ -267,6 +266,17 @@ def send_order_confirmation_email(order: Order, current_site) -> None:
...
+    return eb_ticket_orders.count() > settings.EARLY_BIRD_ORDER_LIMIT
```
Comment (matrixise) [inline conference/cart.py]: `settings.EARLY_BIRD_ORDER_LIMIT` ? I thought this was supposed to be configurable via Fares.

**Review by matrixise — APPROVED** [review]
